server/grpc/runner/runner.go

```diff
@@ -0,0 +1,284 @@
+package runner
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/mesg-foundation/engine/cosmos"
+	"github.com/mesg-foundation/engine/event/publisher"
+	"github.com/mesg-foundation/engine/execution"
+	"github.com/mesg-foundation/engine/hash"
+	"github.com/mesg-foundation/engine/instance"
+	"github.com/mesg-foundation/engine/protobuf/acknowledgement"
+	"github.com/mesg-foundation/engine/runner"
+	executionmodule "github.com/mesg-foundation/engine/x/execution"
+	runnermodule "github.com/mesg-foundation/engine/x/runner"
+	"google.golang.org/grpc/metadata"
+)
+
+// Server is the type to aggregate all Runner APIs.
+type Server struct {
+	rpc *cosmos.RPC
+	ep  *publisher.EventPublisher
+
+	tokenToRunnerHash sync.Map
+}
+
+// NewServer creates a new Server.
+func NewServer(rpc *cosmos.RPC, ep *publisher.EventPublisher) *Server {
+	return &Server{
+		rpc: rpc,
+		ep:  ep,
+	}
+}
+
+// Register register a new runner.
+func (s *Server) Register(ctx context.Context, req *RegisterRequest) (*RegisterResponse, error) {
+	// decode msg
+	var msg runnermodule.MsgCreate
+	if err := s.rpc.Codec().UnmarshalJSON([]byte(req.Msg), &msg); err != nil {
+		return nil, err
+	}
+	if err := msg.ValidateBasic(); err != nil {
+		return nil, err
+	}
+
+	// decode signature
+	signature, err := hex.DecodeString(req.Signature)
+	if err != nil {
+		return nil, err
+	}
+
+	// check signature
+	acc, err := s.rpc.GetAccount()
+	if err != nil {
+		return nil, err
+	}
+	if !acc.GetPubKey().VerifyBytes([]byte(req.Msg), signature) {
+		return nil, fmt.Errorf("verification of the signature failed, it should be signed by %q", acc.GetAddress())
+	}
+
+	// calculate runner hash
+	inst, err := instance.New(msg.ServiceHash, msg.EnvHash)
+	if err != nil {
+		return nil, err
+	}
+	run, err := runner.New(msg.Owner.String(), inst.Hash)
+	if err != nil {
+		return nil, err
+	}
+	runnerHash := run.Hash
+
+	// check that runner doesn't already exist
+	runnerExist := true
+	route := fmt.Sprintf("custom/%s/%s/%s", runnermodule.QuerierRoute, runnermodule.QueryGet, runnerHash)
+	if _, _, err := s.rpc.QueryWithData(route, nil); err != nil {
+		if strings.Contains(err.Error(), fmt.Sprintf("runner %q not found", runnerHash)) {
```

Comment:
i don't like to catch an error but I couldn't find a better way with the current system
maybe it will be nicer with a dedicated error instead of using the error message?

Reply:
or just an endpoint "exists" maybe

Reply:
done in https://github.com/mesg-foundation/engine/pull/1768 and rebased current PR